common/authorization/interceptor.go

```diff
@@ -68,183 +68,217 @@ var (
-		var namespace string
-		requestWithNamespace, ok := req.(hasNamespace)
-		if ok {
-			namespace = requestWithNamespace.GetNamespace()
-		}
-
-		handler := a.getMetricsHandler(metrics.AuthorizationScope, namespace)
-		result, err := a.authorize(ctx, claims, &CallTarget{
-			Namespace: namespace,
-			APIName:   info.FullMethod,
-			Request:   req,
-		}, handler)
-		if err != nil {
-			handler.Counter(metrics.ServiceErrAuthorizeFailedCounter.GetMetricName()).Record(1)
-			a.logAuthError(err)
-			return nil, errUnauthorized // return a generic error to the caller without disclosing details
-		}
-		if result.Decision != DecisionAllow {
-			handler.Counter(metrics.ServiceErrUnauthorizedCounter.GetMetricName()).Record(1)
-			// if a reason is included in the result, include it in the error message
-			if result.Reason != "" {
-				return nil, serviceerror.NewPermissionDenied(RequestUnauthorized, result.Reason)
-			}
-			return nil, errUnauthorized // return a generic error to the caller without disclosing details
-		}
+// PeerCert extracts an x509 certificate from given tlsInfo.
+func PeerCert(tlsInfo *credentials.TLSInfo) *x509.Certificate {
+	if tlsInfo == nil || len(tlsInfo.State.VerifiedChains) == 0 || len(tlsInfo.State.VerifiedChains[0]) == 0 {
+		return nil
 	}
-	return handler(ctx, req)
+	// The assumption here is that we only expect a single verified chain of certs (first[0]).
+	// It's unclear how we should handle a situation when more than one chain is presented,
+	// which subject to use. It's okay for us to limit ourselves to one chain.
+	// We can always extend this logic later.
+	// We take the first element in the chain ([0]) because that's the client cert
+	// (at the beginning of the chain), not intermediary CAs or the root CA (at the end of the chain).
+	return tlsInfo.State.VerifiedChains[0][0]
 }
 
-func (a *interceptor) authorize(
-	ctx context.Context,
-	claims *Claims,
-	callTarget *CallTarget,
-	metricsHandler metrics.Handler) (Result, error) {
-	startTime := time.Now().UTC()
-	defer func() {
-		metricsHandler.Timer(metrics.ServiceAuthorizationLatency.GetMetricName()).Record(time.Since(startTime))
-	}()
-	return a.authorizer.Authorize(ctx, claims, callTarget)
+// HeaderGetter is an interface for getting a single header value from a case insensitive key.
+type HeaderGetter interface {
+	Get(string) string
 }
 
-func (a *interceptor) logAuthError(err error) {
-	a.logger.Error("Authorization error", tag.Error(err))
+// Wrapper for gRPC metadata that exposes a helper to extract a single metadata value.
+type grpcHeaderGetter struct {
+	metadata metadata.MD
 }
 
-type interceptor struct {
-	authorizer          Authorizer
+// Get a single value from the underlying gRPC metadata.
+// Returns an empty string if the metadata key is unset.
+func (h grpcHeaderGetter) Get(key string) string {
+	values := h.metadata.Get(key)
+	if len(values) == 0 {
+		return ""
+	}
+	return values[0]
+}
+
+type Interceptor struct {
```

Comment:
I think we should keep `Authorization` in the struct and constructor names to keep it consistent with the naming convention of our other interceptors.

Reply:
This is an anti-pattern in Go.

Reply:
Do you happen to have any documentation about that philosophy handy? IMO it's better to have our code consistent even if it is an anti-pattern, but I don't feel that strongly if you would rather keep it as is.

Reply:
You can look at the Go standard library for reference.